src/typecheck.rs

```diff
@@ -48,20 +115,63 @@ impl<T: std::fmt::Display> std::fmt::Display for Atom<T> {
     }
 }
 
-#[derive(Default, Debug, Clone)]
-pub struct Query {
-    pub atoms: Vec<Atom<Symbol>>,
-    pub filters: Vec<Atom<Primitive>>,
+#[derive(Debug, Clone)]
+pub struct Query<T> {
+    pub atoms: Vec<Atom<T>>,
```

Comment:
What happened to filters? Wasn't it convenient to separate primitives from atoms?